### DISCUSSION TITLE
Add arm64 CI support

### DISCUSSION THREAD

**Discussion by xin3liang** [rook/rook#7710] [General]
Hi Rook!  This is @xin3liang from [Linaro](https://www.linaro.org/) cloud team.  We focus on improving open source software support for the Arm64 platform.
We notice that Rook is a great cloud-native storage orchestrator project. And Rook is releasing Arm64 container images. This is really awesome!
To make Arm64 as the main platform for Rook which may mean every PR and binary release are tested on the Arm64 platform before they are merged and released. We are interested in adding CI support for Arm64.
Please let me know if you have any ideas and comments on this. Thank you!

**Comment by leseb** [comment]
Hi @xin3liang, thanks for reaching out! As part of our CI, we are using Github Actions quite extensively. I know for a fact that Github Actions does not offer arm64 machine. This means that we would have to [add a self-hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/adding-self-hosted-runners). Is this something you do already?

Thanks!

**Reply by xin3liang** [reply]
Hi @leseb, thanks for the quick reply 👍 . We have a public [Arm64 developer cloud](https://www.linaro.cloud/) for open source projects. We can add a self-hosted runner for the CI.

**Reply by leseb** [reply]
Can you describe the process to follow in order to have a self-hosted runner donated?

**Reply by xin3liang** [reply]
Sure, I will investigate the whole process first according to [add a self-hosted runner](https://docs.github.com/en/actions/hosting-your-own-runners/adding-self-hosted-runners). Thanks @leseb.

**Reply by xin3liang** [reply]
Hi @leseb, 
I have a try to add self-hosted runner and run hello world CI job on my own repository.
I need your help to add self-hosted runners to rook/rook repository. What's your time zone? I will ping you on the Rook slack, ok?
See AArch64 running log: https://github.com/xin3liang/home-bin/runs/2606799343?check_suite_focus=true#step:3:4

**Reply by subhamkrai** [reply]
@xin3liang one question in self-hosted runner we do we get a clean machine with every run?

**Reply by xin3liang** [reply]
> Can you describe the process to follow in order to have a self-hosted runner donated?

The whole process is described here: https://docs.github.com/en/actions/hosting-your-own-runners/adding-self-hosted-runners#adding-a-self-hosted-runner-to-a-repository

During step 6, you will find bellow commands to be run in the self-hosted runner
```
# Create a folder
$ mkdir actions-runner && cd actions-runner# Download the latest runner package
$ curl -o actions-runner-linux-arm64-2.278.0.tar.gz -L https://github.com/actions/runner/releases/download/v2.278.0/actions-runner-linux-arm64-2.278.0.tar.gz# Extract the installer
$ tar xzf ./actions-runner-linux-arm64-2.278.0.tar.gz

# Create the runner and start the configuration experience
$ ./config.sh --url https://github.com/rook/rook --token xxxxxxxxxxx 
# Last step, run it!
$ ./run.sh

# Use this YAML in your workflow file for each job
runs-on: self-hosted
```

**Reply by xin3liang** [reply]
> @xin3liang one question in self-hosted runner we do we get a clean machine with every run?

The self-hosted runners are dedicated hosts/VMs.  So it's not a new created and clean machine with every run. But actions will handle the clean up for each time it runs, I think. 
See here for more differences between github-hosted and selft-hosted runners: https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#differences-between-github-hosted-and-self-hosted-runners

**Reply by leseb** [reply]
> Hi @leseb,
> I have a try to add self-hosted runner and run hello world CI job on my own repository.
> I need your help to add self-hosted runners to rook/rook repository. What's your time zone? I will ping you on the Rook slack, ok?
> See AArch64 running log: https://github.com/xin3liang/home-bin/runs/2606799343?check_suite_focus=true#step:3:4

Awesome! I'm CEST — Central European Summer Time, feel free to ping me on slack!

**Reply by leseb** [reply]
I have prepared the config on the Rook repo, ping me on Slack so I can give you the token :)

**Reply by leseb** [reply]
@xin3liang ^

**Comment by xin3liang** [comment]
Thanks @leseb 's help. I succeed to add two AArch64 self-hosted ubuntu-20.04 runners now. Here are the cmds I run in the self-hosted runners :
```
## Add self-hosted runner and start service
TOKEN=xxxxxx
REPO=https://github.com/rook/rook 
./config.sh --url $REPO --token $TOKEN --labels ubuntu-20.04 --name $(hostname) --work _work
sudo ./svc.sh install
sudo ./svc.sh start
sudo ./svc.sh status
```

**Reply by leseb** [reply]
PR https://github.com/rook/rook/pull/7946

**Comment by leseb** [comment]
The job is successfully running now, scheduled every day, see: https://github.com/rook/rook/actions/workflows/canary-integration-test-arm64.yml?query=event%3Aschedule

Thanks again @xin3liang

**Reply by xin3liang** [reply]
> The job is successfully running now, scheduled every day, see: https://github.com/rook/rook/actions/workflows/canary-integration-test-arm64.yml?query=event%3Aschedule
> 
> Thanks again @xin3liang

Well done! Thanks  @leseb .
